Fix non-msvc compiler warnings

Biswa96 opened the pull request.

hinglau886 commented on absl/base/internal/direct_mmap.h:
```diff
@@ -20,7 +20,7 @@
 
 #include "absl/base/config.h"
 
-#if ABSL_HAVE_MMAP
```
absl/base/internal/direct_mmap.h

Biswa96 committed:
Check _MSC_VER for MSVC specific pragma directives

This fixes the following warnings with mingw toolchain
warning: ignoring '#pragma comment ' [-Wunknown-pragmas]
warning: ignoring '#pragma warning ' [-Wunknown-pragmas]

Biswa96 committed:
Fix some preprocessor evaluation

This fixes the following warnings with mingw toolchain
warning: __clang_major__ is not defined, evaluates to 0 [-Wundef]
warning: HAS_STRPTIME is not defined, evaluates to 0 [-Wundef]
warning: ABSL_HAVE_MMAP is not defined, evaluates to 0 [-Wundef]

Biswa96 committed:
Include pthread headers for mingw

Biswa96 commented:
I have rebased the commits to latest master branch. Please review this pull request.
